### PR TITLE
feat(babel-plugin): initial version

### PR DESCRIPTION
**Pull request by tatchi**
My attempt to implement #10. Don't be afraid by all the files, lots of them are actually tests 😅

Don't hesitate if you have any comments, that's the first time I create a babel plugin so there are chances things are not implemented in an idiomatic way.

Let me know also from the tests if something is missing/not working properly. There's still at least one thing that I'm aware of which is not working: if one has already imported the hook using an alias

```js
import { useStyling as useCustomName } from `glaze`
```

The plugin will detect that the hook is already imported but won't use the alias to call the hook:

```js
const sx = useStyling(); // should be useCustomName()
```
There's a test `importAlias` to should highlight that.

### CODE REVIEW

File: .eslintignore [hunk 1]
```diff
@@ -0,0 +1 @@
+**/babel-plugin-glaze/__test__/**
```
Comment (tatchi) [inline .eslintignore]: I had a lot of eslint errors when trying to commit. Turned out that lots of them were coming from the test files. Don't know If it's the best way to ignore them but at least it worked :P

File: packages/babel-plugin-glaze/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,305 @@
+/* eslint-disable @typescript-eslint/explicit-function-return-type */
+/* eslint-disable no-param-reassign */
+import { NodePath, PluginObj, template, types } from '@babel/core';
```
Comment (tatchi) [inline packages/babel-plugin-glaze/src/index.ts:3]: I had to turn off some of the eslint lint for this file. Was a bit lazy to properly handle them and actually I'm not sure if that makes sense to enable them for this file...

File: packages/babel-plugin-glaze/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,305 @@
+/* eslint-disable @typescript-eslint/explicit-function-return-type */
+/* eslint-disable no-param-reassign */
```
Comment (tatchi) [inline packages/babel-plugin-glaze/src/index.ts]: I had to turn off some of the eslint lint for this file. Was a bit lazy to properly handle them and actually I'm not sure if that makes sense to enable them for this file...

**Comment by tatchi** [comment]
I fixed some of the eslint errors. Normally the remaining ones are due to the test files. 

Do you know how could I disable the check for these files?

File: .eslintignore (hunk 1 above)
Comment (kripod) [reply .eslintignore]: ESLint is [configured](https://github.com/kripod/glaze/blob/08cc3fd809df7db2af68767f2c618a109b2a523c/package.json#L12) to only ignore the files specified in `.gitignore` Unfortunately, the `.eslintignore` file will have no effect.

File: packages/babel-plugin-glaze/src/index.ts (hunk 3 above)
Comment (kripod) [reply packages/babel-plugin-glaze/src/index.ts]: I'll check whether enabling them is worth it. I prefer enabling/disabling rules on a per-line or per-project basis. The latter can be done through `.eslintrc`.

Comment (kripod) [reply packages/babel-plugin-glaze/src/index.ts]: Also, thank you for highlighting this! 😊

**Comment by kripod** [comment]
Thank you very much for all the efforts made towards making this feature available! I'll try my best to review the code as soon as possible, trying to fix some issues while learning Babel myself 😄

**Comment by kripod** [comment]
Currently, I'm working on multi-page docs before releasing v0.5. After that, I'll be happy to inspect and review these wonderful contributions. In the meantime, I thought about adding proper autocomplete for TypeScript and had the idea that the `sx` prop could be allowed by:

```ts
/* 'treat.d.ts', placed in projects which depend on glaze */

import 'glaze/sx-prop'; // A '.d.ts' file in the glaze package root


// Just the usual from below, as seen in 'packages/example-gatsby':

import { tokens } from './theme.treat';

declare module 'treat/theme' {
  type Tokens = typeof tokens;
  // eslint-disable-next-line @typescript-eslint/no-empty-interface
  export interface Theme extends Tokens {}
}
```

Implementation ideas for `glaze/sx-prop.d.ts`:

- [Styled Components](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/72c1e33e13affa5fa81bd1460eb8881e0d97ddca/types/styled-components/cssprop.d.ts) – **preferred,** as [`interface IntrinsicAttributes extends React.Attributes`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/72c1e33e13affa5fa81bd1460eb8881e0d97ddca/types/react/index.d.ts#L2962)
- [Emotion](https://github.com/emotion-js/emotion/blob/9776758ad9beac552c4b8fe535beaaa62e1af6a1/packages/core/types/index.d.ts#L82-L99), [Theme UI](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/72c1e33e13affa5fa81bd1460eb8881e0d97ddca/types/theme-ui/index.d.ts#L223-L233)

The recommended `.d.ts` file should be documented, maybe on a page dedicated to _usage with TypeScript._

**Comment by kripod** [comment]
Quite possibly, the `useStyling` hook's implementation will be split into 2 parts:

1. Static style resolver, pairing CSS objects to the static class names generated by treat
2. Dynamic style injector, as an optional fallback

The underlying [`ThemedStyle`](https://github.com/kripod/glaze/blob/2c4edc2163ebb1bad04f46872361bc999a2d46f5/packages/glaze/src/useStyling.ts#L32-L48) type should be reused for the `sx` prop's typings, but it will possibly be moved to a new location.

**Comment by kripod** [comment]
Fantastic work, thank you so much! 🙌

I would appreciate if you could document the plugin's usage and [known caveat](https://github.com/tatchi/glaze/blob/a4680ef9474594bed19a8df5a5078b672d90aead/packages/babel-plugin-glaze/tests/fixtures/import-alias/output.jsx#L2) inside the readme, linking it to #27. Please let me know if you need any kind of assistance!

**Review by kripod — APPROVED** [review]

**Comment by kripod** [comment]
@all-contributors please add @tatchi for code and tests

**Comment by kripod** [comment]
@all-contributors please add @tatchi for code and tests

**Comment by kripod** [comment]
I got an even better idea for adding the `sx` prop types to elements: Create a package called `@types/babel-plugin-glaze` in [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped) and then add it as a dependency of `babel-plugin-glaze` to auto-import the `.d.ts` file.

_(The work has started at https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44048)_

**Comment by kripod** [comment]
The [type definitions](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44048) are now ready to roll, awaiting to be merged into the DefinitelyTyped repository. Once they are ready, adding `@types/babel-plugin-glaze` to the `dependencies` field of `babel-plugin-glaze` will make typings available for the `sx` prop.

**Pull request merged**